src/sentry/snuba/query_subscription_consumer.py

```diff
@@ -0,0 +1,166 @@
+        self.initial_offset_reset = initial_offset_reset
+
+    def run(self):
+        logger.debug('Starting snuba query subscriber')
+
+        conf = {
+            'bootstrap.servers': self.bootstrap_servers,
+            'group.id': self.group_id,
+            'session.timeout.ms': 6000,
+            'auto.offset.reset': self.initial_offset_reset,
+            'enable.auto.commit': 'false',
+            'enable.auto.offset.store': 'true',
+            'enable.partition.eof': 'false',
+            'default.topic.config': {'auto.offset.reset': self.initial_offset_reset},
+        }
+        consumer = Consumer(conf)
+        consumer.subscribe([self.topic])
+        try:
+            i = 0
+            while True:
+                message = consumer.poll(0.1)
+                if message is None:
+                    continue
+
+                error = message.error()
+                if error is not None:
+                    raise KafkaException(error)
+
+                i = i + 1
+
+                self.handle_message(message)
+
+                if i % self.commit_batch_size == 0:
+                    logger.debug('Committing offsets')
+                    consumer.commit()
+        except KeyboardInterrupt:
+            pass
+        logger.debug('Committing offsets and closing consumer')
+        consumer.commit()
+        consumer.close()
+
+    def handle_message(self, message):
+        """
+        Parses the value from Kafka, and if valid passes the payload to the callback defined by the
+        subscription. If the subscription has been removed, or no longer has a valid callback then
+        just log metrics/errors and continue.
+        :param message:
+        :return:
+        """
+        try:
+            contents = self.parse_message_value(message.value())
+        except InvalidMessageError:
+            # If the message is in an invalid format, just log the error
+            # and continue
+            logger.exception('Subscription update could not be parsed')
+            return
+
+        try:
+            subscription = QuerySubscription.objects.get(
+                subscription_id=contents['subscription_id'])
+        except QuerySubscription.DoesNotExist:
+            metrics.incr('snuba_query_subscriber.subscription_doesnt_exist')
+            logger.error('Received subscription update, but subscription does not exist')
+            return
+
+        if subscription.type not in subscriber_registry:
+            metrics.incr('snuba_query_subscriber.subscription_type_not_registered')
+            logger.error('Received subscription update, but no subscription handler registered')
+            return
+
+        callback = subscriber_registry[subscription.type]
+        with metrics.timer(
+            'snuba_query_subscriber.callback.duration',
+            instance=subscription.type,
+        ):
+            callback(contents, subscription)
+
+    def parse_message_value(self, value):
```

Comment:
Potentially we should convert this to an intermediate format, rather than just passing the validated json around. Not sure if super important?